lib/stubs/object_ml.c

```diff
@@ -0,0 +1,135 @@
+#include "lablqml.h"
+#include "variant.h"
+
+#include <caml/memory.h>
+#include <caml/threads.h>
+#include <caml/alloc.h>
+#include <caml/callback.h>
+#include <caml/custom.h>
+#include <assert.h>
+
+#include "object.h"
+
+extern "C" void ocamlobject_callback (value *ocaml_function, QVariant v);
+
+OCamlBinding::OCamlBinding(QObject *obj, const QString &name, value *func):
+QObject(obj), ocaml_function(func), property(obj, name)
+{
+  property.connectNotifySignal(this, SLOT(valueChanged()));
+}
+
+OCamlBinding::~OCamlBinding(){
+  if (ocaml_function != NULL) {
+    caml_acquire_runtime_system();
+    caml_remove_global_root(ocaml_function);
+    caml_release_runtime_system();
+    free(ocaml_function);
+  }
+  qDebug () << "destroyed object for " << property.name();
+}
+
+bool OCamlBinding::write(QVariant v) {
+  return property.write(v);
+}
+
+void OCamlBinding::valueChanged () {
+  if (ocaml_function != NULL) {
+    caml_acquire_runtime_system();
+    ocamlobject_callback (ocaml_function, property.read ());
+    caml_release_runtime_system();
+  }
+}
+
+OCamlObject::OCamlObject(QObject *parent):
+QObject(parent)
+{
+  caml_c_thread_register();
+}
+
+OCamlObject::~OCamlObject(){
+  caml_c_thread_unregister();
+  qDebug () << "object destroyed";
+}
+
+bool OCamlObject::write(QQmlProperty property, QVariant v) {
+  return property.write(v);
+}
+
+extern "C" {
+
+  void free_ocaml_binding_object(value v){
+    QMetaObject::invokeMethod (Ctype_of_val(OCamlBinding, v), "deleteLater", Qt::QueuedConnection);
+  }
+
+  static struct custom_operations ops = {
+    "lablqml.qml.property",
+    free_ocaml_binding_object,
+    custom_compare_default,
+    custom_hash_default,
+    custom_serialize_default,
+    custom_deserialize_default,
+    custom_compare_ext_default
+  };
+
+  value ocamlobject_binding(value create, value qt_object_val, value property_name_val, value func_val) {
+    CAMLparam4(create, qt_object_val, property_name_val, func_val);
+    CAMLlocal1(result_val);
+
+    QObject *o = Ctype_field(QObject, qt_object_val, 0);
+    Q_ASSERT(o != nullptr);
+
+    if (!Bool_val(create)) {
+      QVariant property_object = o->property(String_val(property_name_val));
+      if (!property_object.isValid())
+	caml_failwith("Property not found");
+    }else{
+      caml_failwith("Implicit creation not supported yet");
+    }
+
+    value *ocaml_function = (value*) malloc(sizeof(func_val));
```

Comment:
I think it's the `free()` which needs to go in `void free_ocaml_binding_object(value v)`, because OCaml might want to destroy the callback function (not supported yet), but the the QObject might not be released until later because of [deleteLater](https://github.com/Kakadu/lablqml/blob/600f470cd89e92e74e6400e2612d37d7b965b01a/lib/stubs/object_ml.c#L61).

From what I understood, `QObject`s are supposed to be owned by the Qt runtime system. So I want to keep the OCaml memory as separated and discrete as possible to avoid assumptions of about owning the callbacks.

What do you think?